File: backend/app/models/agent.py
"""Agent mode request and response models."""

from typing import Any, Literal

from pydantic import BaseModel, Field


# === Request Models ===


class AgentQueryRequest(BaseModel):
    """Agent query request."""

    prompt: str = Field(
        ...,
        min_length=1,
        max_length=4000,
        description="User's natural language request",
    )


# === Event Models ===


class ThinkingEventData(BaseModel):
    """Thinking status event data."""

    status: Literal["analyzing", "planning", "generating"]
    message: str


class ToolCallEventData(BaseModel):
    """Tool call event data."""

    id: str
    tool: str
    input: dict[str, Any]
    status: Literal["running", "completed", "error"]
    output: str | None = None
    duration_ms: int | None = None


class MessageEventData(BaseModel):
    """Assistant message event data."""

    role: Literal["assistant"] = "assistant"
    content: str


class SQLEventData(BaseModel):
    """Final SQL event data."""

    sql: str
    explanation: str | None = None


class ErrorEventData(BaseModel):
    """Error event data."""

    error: str
    detail: str | None = None


class DoneEventData(BaseModel):
    """Done event data."""

    total_time_ms: int
    tool_calls_count: int


# === Response Models ===


class AgentStatusResponse(BaseModel):
    """Agent status response."""

    available: bool
    configured: bool
    model: str | None = None


class CancelResponse(BaseModel):
    """Cancel response."""

    cancelled: bool

File: frontend/src/types/agent.ts
// === Agent Message Types ===

export type AgentMessageRole = 'user' | 'assistant' | 'tool';

export interface AgentMessage {
  id: string;
  role: AgentMessageRole;
  content: string;
  timestamp: number;
  toolCall?: ToolCallInfo;
}

export interface ToolCallInfo {
  id: string;
  name: string;
  input: Record<string, unknown>;
  output?: string;
  status: 'running' | 'completed' | 'error';
  durationMs?: number;
}

// === SSE Event Types ===

export type AgentEventType =
  | 'thinking'
  | 'tool_call'
  | 'tool_result'
  | 'message'
  | 'sql'
  | 'error'
  | 'done';

export interface ThinkingEventData {
  status: 'analyzing' | 'planning' | 'generating';
  message: string;
}

export interface ToolCallEventData {
  id: string;
  tool: string;
  input: Record<string, unknown>;
  status: 'running' | 'completed' | 'error';
  output?: string;
  durationMs?: number;
}

export interface MessageEventData {
  role: 'assistant';
  content: string;
}

export interface SQLEventData {
  sql: string;
  explanation?: string;
}

export interface ErrorEventData {
  error: string;
  detail?: string;
}

export interface DoneEventData {
  totalTimeMs: number;
  toolCallsCount: number;
}

// === Component State Types ===

export type AgentStatus =
  | 'idle'
  | 'thinking'
  | 'tool_running'
  | 'responding'
  | 'completed'
  | 'error'
  | 'cancelled';

export interface AgentState {
  status: AgentStatus;
  messages: AgentMessage[];
  generatedSQL?: string;
  explanation?: string;
  error?: string;
  totalTimeMs?: number;
}

// === API Types ===

export interface AgentQueryRequest {
  prompt: string;
}

export interface AgentStatusResponse {
  available: boolean;
  configured: boolean;
  model?: string;
}

// === SSE Event Handler Types ===

export interface AgentEventHandlers {
  onThinking?: (data: ThinkingEventData) => void;
  onToolCall?: (data: ToolCallEventData) => void;
  onMessage?: (data: MessageEventData) => void;
  onSQL?: (data: SQLEventData) => void;
  onError?: (data: ErrorEventData) => void;
  onDone?: (data: DoneEventData) => void;
}

